alibi_detect/cd/cvm_online.py

```diff
@@ -0,0 +1,267 @@
+                stats[k, :] = _normalise_stats(sum_diffs_2, self.n, ws)
+            else:
+                stats[k, :] = np.nan
+        return stats
+
+    def _check_drift(self, test_stats: np.ndarray, thresholds: np.ndarray) -> int:
+        """
+        Private method to compare test stats to thresholds. The max stats over all windows are compute for each
+        feature. Drift is flagged if `max_stats` for any feature exceeds the single `thresholds` set.
+
+        Parameters
+        ----------
+        test_stats
+            Array of test statistics with shape (n_windows, n_features)
+        thresholds
+            Array of thresholds with shape (t_max, 1).
+
+        Returns
+        -------
+        An int equal to 1 if drift, 0 otherwise.
+        """
+        with warnings.catch_warnings():
+            warnings.filterwarnings(action='ignore', message='All-NaN slice encountered')
+            max_stats = np.nanmax(test_stats, axis=0)
+        drift_pred = int((max_stats > thresholds).any())
+        return drift_pred
+
+
+@nb.njit(parallel=False, cache=True)
+def _normalise_stats(stats: np.ndarray, n: int, ws: int) -> np.ndarray:
+    """
+    See Eqns 3 & 14 of https://www.projecteuclid.org/euclid.aoms/1177704477.
+    """
+    mu = 1 / 6 + 1 / (6 * (n + ws))
+    var_num = (n + ws + 1) * (4 * n * ws * (n + ws) - 3 * (n * n + ws * ws) - 2 * n * ws)
+    var_denom = 45 * (n + ws) * (n + ws) * 4 * n * ws
+    prod = n * ws / ((n + ws) * (n + ws))
+    return (stats * prod - mu) / np.sqrt(var_num / var_denom)
+
+
+@nb.njit(parallel=True, cache=True)
+def _ids_to_stats(
+        ids_ref_all: np.ndarray,
+        ids_stream_all: np.ndarray,
+        window_sizes: np.ndarray
+) -> np.ndarray:
+    n_bootstraps = ids_ref_all.shape[0]
+    n = ids_ref_all.shape[1]
+    t_max = ids_stream_all.shape[1]
+    n_all = ids_ref_all.shape[-1]
+    n_windows = window_sizes.shape[0]
+
+    stats = np.zeros((n_bootstraps, t_max, n_windows))
+
+    for b in nb.prange(n_bootstraps):
+        ref_cdf_all = np.sum(ids_ref_all[b], axis=0) / n
+
+        cumsums = np.zeros((t_max+1, n_all))
+        for i in range(n_all):
+            cumsums[1:, i] = np.cumsum(ids_stream_all[b, :, i])
+
+        for k in range(n_windows):
+            ws = window_sizes[k]
+            win_cdf_ref = (cumsums[ws:, :n] - cumsums[:-ws, :n]) / ws
+            cdf_diffs_on_ref = np.empty_like(win_cdf_ref)
+            for j in range(win_cdf_ref.shape[0]):  # Need to loop through as can't broadcast in njit parallel
+                cdf_diffs_on_ref[j, :] = ref_cdf_all[:n] - win_cdf_ref[j, :]
+            stats[b, (ws-1):, k] = np.sum(cdf_diffs_on_ref * cdf_diffs_on_ref, axis=-1)
+            for t in range(ws, t_max):
+                win_cdf_win = (cumsums[t, n + t - ws:n + t] - cumsums[t - ws, n + t - ws:n + t]) / ws
+                cdf_diffs_on_win = ref_cdf_all[n + t - ws:n + t] - win_cdf_win
+                stats[b, t, k] += np.sum(cdf_diffs_on_win * cdf_diffs_on_win)
```

Comment:
Sadly I think there's still out-by-1 errors in this for loop. I think it should start from ws-1 and we need to add 1 to ts indexing cumsums.